lib/handlers/pools/provider-migration/v3/traffic-switch-v3-pool-provider.ts

```diff
@@ -0,0 +1,120 @@
+
+  protected readonly SHOULD_SWITCH_TRAFFIC = () =>
+    POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.switchPercentage > this.getRandomPercentage()
+  protected readonly SHOULD_SAMPLE_TRAFFIC = () =>
+    POOL_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.samplingPercentage > this.getRandomPercentage()
+
+  constructor({ currentPoolProvider, targetPoolProvider, sourceOfTruthPoolProvider }: TrafficSwitchPoolProviderProps) {
+    this.currentPoolProvider = currentPoolProvider
+    this.targetPoolProvider = targetPoolProvider
+    this.sourceOfTruthPoolProvider = sourceOfTruthPoolProvider
+  }
+
+  getPoolAddress(
+    tokenA: Token,
+    tokenB: Token,
+    feeAmount: FeeAmount
+  ): {
+    poolAddress: string
+    token0: Token
+    token1: Token
+  } {
+    // The underlying logic for getting the pool address is always from the
+    // source of truth pool provider, and the result is deterministic.
+    // It doesn't matter which pool provider we use, so we can start with the target
+    // pool provider, which will delegate to the source of truth pool provider.
+    return this.targetPoolProvider.getPoolAddress(tokenA, tokenB, feeAmount)
+  }
+
+  async getPools(tokenPairs: [Token, Token, FeeAmount][], providerConfig?: ProviderConfig): Promise<V3PoolAccessor> {
+    const currentProviderPools = await this.currentPoolProvider.getPools(tokenPairs, providerConfig)
+
+    metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_TOTAL', 1, MetricLoggerUnit.None)
+    const sampleTraffic = this.SHOULD_SAMPLE_TRAFFIC()
+    if (sampleTraffic) {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_SAMPLING', 1, MetricLoggerUnit.None)
+      const targetProviderPools = await this.targetPoolProvider.getPools(tokenPairs, providerConfig)
+
+      // If we need to sample the traffic, we don't want to make it a blocking I/O
+      this.sampleTraffic(tokenPairs, currentProviderPools, targetProviderPools, providerConfig)
+    }
+
+    const switchTraffic = this.SHOULD_SWITCH_TRAFFIC()
+    if (!switchTraffic) {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_CURRENT', 1, MetricLoggerUnit.None)
+      return currentProviderPools
+    } else {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_TARGET', 1, MetricLoggerUnit.None)
+      return await this.targetPoolProvider.getPools(tokenPairs, providerConfig)
+    }
+  }
+
+  private async sampleTraffic(
+    tokenPairs: [Token, Token, FeeAmount][],
+    currentProviderPools: V3PoolAccessor,
+    targetProviderPools: V3PoolAccessor,
+    providerConfig?: ProviderConfig
+  ): Promise<void> {
+    const truthProviderPools = await this.sourceOfTruthPoolProvider.getPools(tokenPairs, providerConfig)
+
+    truthProviderPools.getAllPools().forEach((pool: Pool) => {
+      const currentProviderPool = currentProviderPools.getPool(pool.token0, pool.token1, pool.fee)
+      if (!currentProviderPool) {
+        // We don't expect missing pool, but export metric in case we see any
```

Comment:
this can happen if there is a cache miss still right?

Reply:
Actually it shouldn't. The reason is because if there's a cache miss from `CachingV3PoolProvider`, then it will delegate to the actual provider `V3PoolProvider` to get the pool.

Reply:
sorry n00b question, but does the delegation happen within the `CachingV3PoolProvider`?

Reply:
Nw, let me copy paste the code block:

```
export class CachingV3PoolProvider implements IV3PoolProvider {
  private POOL_KEY = (chainId: ChainId, address: string) =>
    `pool-${chainId}-${address}`;

  /**
   * Creates an instance of CachingV3PoolProvider.
   * @param chainId The chain id to use.
   * @param poolProvider The provider to use to get the pools when not in the cache.
   * @param cache Cache instance to hold cached pools.
   */
  constructor(
    protected chainId: ChainId,
    protected poolProvider: IV3PoolProvider,
    private cache: ICache<Pool>
  ) {}

  public async getPools(
    tokenPairs: [Token, Token, FeeAmount][],
    providerConfig?: ProviderConfig
  ): Promise<V3PoolAccessor> {
    const poolAddressSet: Set<string> = new Set<string>();
    const poolsToGetTokenPairs: Array<[Token, Token, FeeAmount]> = [];
    const poolsToGetAddresses: string[] = [];
    const poolAddressToPool: { [poolAddress: string]: Pool } = {};

    for (const [tokenA, tokenB, feeAmount] of tokenPairs) {
      const { poolAddress, token0, token1 } = this.getPoolAddress(
        tokenA,
        tokenB,
        feeAmount
      );

      if (poolAddressSet.has(poolAddress)) {
        continue;
      }

      poolAddressSet.add(poolAddress);

      const cachedPool = await this.cache.get(
        this.POOL_KEY(this.chainId, poolAddress)
      );
      if (cachedPool) {
        poolAddressToPool[poolAddress] = cachedPool;
        continue;
      }

      poolsToGetTokenPairs.push([token0, token1, feeAmount]);
      poolsToGetAddresses.push(poolAddress);
    }

    log.info(
      {
        poolsFound: _.map(
          Object.values(poolAddressToPool),
          (p) => `${p.token0.symbol} ${p.token1.symbol} ${p.fee}`
        ),
        poolsToGetTokenPairs: _.map(
          poolsToGetTokenPairs,
          (t) => `${t[0].symbol} ${t[1].symbol} ${t[2]}`
        ),
      },
      `Found ${
        Object.keys(poolAddressToPool).length
      } V3 pools already in local cache. About to get liquidity and slot0s for ${
        poolsToGetTokenPairs.length
      } pools.`
    );

    if (poolsToGetAddresses.length > 0) {
      const poolAccessor = await this.poolProvider.getPools(
        poolsToGetTokenPairs,
        providerConfig
      );
      for (const address of poolsToGetAddresses) {
        const pool = poolAccessor.getPoolByAddress(address);
        if (pool) {
          poolAddressToPool[address] = pool;
          await this.cache.set(this.POOL_KEY(this.chainId, address), pool);
        }
      }
    }
```